pkg/download/goget/goget.go

```diff
@@ -0,0 +1,173 @@
+	gomodContent := []byte("module mod")
+	if err := afero.WriteFile(fs, gomodPath, gomodContent, 0666); err != nil {
+		return errors.E(op, err)
+	}
+	sourcePath := filepath.Join(repoRoot, "mod.go")
+	sourceContent := []byte(`package mod`)
+	if err := afero.WriteFile(fs, sourcePath, sourceContent, 0666); err != nil {
+		return errors.E(op, err)
+	}
+	return nil
+}
+
+func (gg *goget) Info(ctx context.Context, mod string, ver string) (*storage.RevInfo, error) {
+	const op errors.Op = "goget.Info"
+	fetcher, _ := module.NewGoGetFetcher(gg.goBinPath, gg.fs) // TODO: remove err from func call
+	ref, err := fetcher.Fetch(mod, ver)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer ref.Clear()
+	v, err := ref.Read()
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer v.Zip.Close()
+
+	var ri storage.RevInfo
+	err = json.Unmarshal(v.Info, &ri)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	return &ri, nil
+}
+
+func (gg *goget) Latest(ctx context.Context, mod string) (*storage.RevInfo, error) {
+	const op errors.Op = "goget.Latest"
+	lr, err := gg.list(op, mod)
+	if err != nil {
+		return nil, err
+	}
+
+	pseudoInfo := strings.Split(lr.Version, "-")
+	if len(pseudoInfo) < 3 {
+		return nil, errors.E(op, fmt.Errorf("malformed pseudoInfo %v", lr.Version))
+	}
+	return &storage.RevInfo{
+		Name:    pseudoInfo[2],
+		Short:   pseudoInfo[2],
+		Time:    lr.Time,
+		Version: lr.Version,
+	}, nil
+}
+
+func (gg *goget) list(op errors.Op, mod string) (*listResp, error) {
+	hackyPath, err := afero.TempDir(gg.fs, "", "hackymod")
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer gg.fs.RemoveAll(hackyPath)
+	err = dummyMod(gg.fs, hackyPath)
+	cmd := exec.Command(
+		gg.goBinPath,
+		"list", "-m", "-versions", "-json",
+		config.FmtModVer(mod, "latest"),
+	)
+	cmd.Dir = hackyPath
+
+	bts, err := cmd.CombinedOutput()
+	if err != nil {
+		errFmt := fmt.Errorf("%v: %s", err, bts)
+		return nil, errors.E(op, errFmt)
+	}
+
+	var lr listResp
+	err = json.Unmarshal(bts, &lr)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	return &lr, nil
+}
+
+func (gg *goget) GoMod(ctx context.Context, mod string, ver string) ([]byte, error) {
+	const op errors.Op = "goget.Info"
+	fetcher, _ := module.NewGoGetFetcher(gg.goBinPath, gg.fs) // TODO: remove err from func call
+	ref, err := fetcher.Fetch(mod, ver)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer ref.Clear()
+	v, err := ref.Read()
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer v.Zip.Close()
+
+	return v.Mod, nil
+}
+
+func (gg *goget) Zip(ctx context.Context, mod string, ver string) (io.ReadCloser, error) {
+	const op errors.Op = "goget.Info"
+	fetcher, _ := module.NewGoGetFetcher(gg.goBinPath, gg.fs) // TODO: remove err from func call
+	ref, err := fetcher.Fetch(mod, ver)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	v, err := ref.Read()
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	return v.Zip, nil
+}
```

Comment:
the code in this file looks very similar to the `./pkg/module/go_get_fetcher.go` and `disk_ref.go` - any way you could build the goget protocol on top of those two primitives?

Reply:
@arschles I am. Line 162 above in this function

Reply:
I take this back completely - I see you are using the Fetcher and Ref!